File: 01. DataStructures/01. Linked List/3. Circular-Linked-List/Circular-Linked-List.cpp
#include <iostream>
using namespace std;

class CircularLinkedList
{
public:
    class Node
    {
    public:
        int data;
        Node *next;

        Node(int d)
        {
            data = d;
            next = NULL;
        }
    };

    int size;
    Node *head;
    Node *tail;

    CircularLinkedList()
    {
        size = 0;
        head=NULL;
        tail =NULL;
    }

//1.Insert node at beginning
    public:
    void InsertNodeFirst(int val)
    {
        Node *newNode = new Node(val);
        if (tail == NULL) //is my LL empty?
        {      
           tail=newNode;
        }
        else
        {
        newNode->next = tail->next;
        }     
       tail->next=newNode;
       size++;
     }

    //2.Insert node at end
    void InsertNodeLast(int val)
    {
        Node *newNode = new Node(val);
        if (tail == NULL)
        {
            tail=newNode;
            tail->next=tail;
        }
        else
        {
           newNode->next=tail->next;
           tail->next=newNode;
           tail=newNode;
        }
        size++;
}

//3. Insert node at pos
    void InsertAtPos(int pos, int val)
    {
        Node *newNode = new Node(val);
        if (tail == NULL)
        {
            cout << "LL is already empty";
            return;
        }
        else if (pos == 1)
        {
            InsertNodeFirst(val);
        }
        else if (pos == size + 1)
        {
            InsertNodeLast(val);
        }
        else
        {
            Node *ptr =tail->next;
            for (int i = 0; i < pos - 2; i++)
            {
                ptr = ptr->next;
            }
            newNode->next = ptr->next;
            ptr->next = newNode;
            size++;
        }
    }

    // 4. Delete node first
    void DeleteNodeFirst()
    {
        if (tail == NULL)
        {
            cout << "LL is empty";
        }
        else
        {
            Node *ptr = tail->next;
            tail->next=ptr->next;
            ptr=ptr->next;
            size--;
        }
    }

    //5. Delete node last
    void DeleteNodeLast()
    {
        if (tail == NULL)
        {
            cout << "LL is empty";
            return;
        }
        else
        {
            Node *ptr = tail->next;
            while (ptr->next != tail)
            {
                ptr = ptr->next;
            }
            ptr->next=tail->next;
            tail=ptr;
            size--;
        }
    }

    //6. Delete node at pos
    void DeleteNodeAtPosition(int pos)
    {
        if (tail == NULL || pos > size || pos < 1)
        {
            cout << "Delete not possible";
            return;
        }
        else if (pos == 1)
        {
            DeleteNodeFirst();
        }
        else if (pos == size)
        {
            DeleteNodeLast();
        }
        else
        {
            Node *ptr = tail->next;
            for (int i = 0; i < pos - 2; i++)
            {
                ptr = ptr->next;
            }
            Node *temp = ptr->next;
            ptr->next = ptr->next->next;
            size--;
        }
    }

    //7. Reverse  list
    void ReverseList()
    {       
            Node* last = tail->next;
            Node *curr = tail->next;
            Node *pre = tail;
            Node* next=NULL;

            while (curr != tail)
            {
                next = curr->next;
                curr->next = pre;
                pre = curr;
                curr = next;
            }
            tail->next=pre;
            tail=last;
    }

//8. Update list
    void UpdateList(int pos, int val)
    {
        if (tail == NULL || pos >size || pos < 1)
        {
            cout << "Update not possible";
            return;
        }
        else
        {
            Node *ptr=tail->next;
            for (int i = 0; i < pos-1; i++)
            {
                ptr = ptr->next;
            }
            ptr->data = val;
        }
    }

    //9. Count/Find list size
    int CountList()
    {
        return size;
    }

    //10. Display list
    void DisplayList()
    {
        if (tail == NULL)
        {
            cout << "No list to Display";
            return;
        }
        Node *ptr = tail->next;
        while (ptr != tail)
        {
            cout << ptr->data << " ---->";
            ptr = ptr->next;
        }
        cout <<ptr->data;
    }
};

int main()
{
    CircularLinkedList Sl;
    int ch;
    int pos, val;
    do
    {
        cout << "\n *******MENU*******\n"
             << endl;
        cout << "1______ Insert Node at First." << endl;
        cout << "2______Insert Node at the Last." << endl;
        cout << "3______Insert Node at any given position." << endl;
        cout << "4______Delete Node at first." << endl;
        cout << "5______Delete Node at last." << endl;
        cout << "6______Delete Node at any given position." << endl;
        cout << "7______Reverse the List." << endl;
        cout << "8______Update a Node." << endl;
        cout << "9______Count the List." << endl;
        cout << "10______Display the List." << endl;
        cout << "0.....EXIT." << endl;
        cout << "\nEnter your choice : ";
        cin >> ch;

        switch (ch)
        {
        case 0:
            cout << "\nEnd of Program" << endl;
            break;
        case 1:
            cout << "Enter value: ";
            cin >> val;
            Sl.InsertNodeFirst(val);
            break;
        case 2:
            cout << "Enter value: ";
            cin >> val;
            Sl.InsertNodeLast(val);
            break;
        case 3:
            cout << "Enter pos: ";
            cin >> pos;
            cout << "Enter value: ";
            cin >> val;
            Sl.InsertAtPos(pos, val);
            break;
        case 4:
            Sl.DeleteNodeFirst();
            break;
        case 5:
            Sl.DeleteNodeLast();
            break;
        case 6:
            cout << "Enter pos: ";
            cin >> pos;
            Sl.DeleteNodeAtPosition(pos);
            break;
        case 7:
            Sl.ReverseList();
            break;
        case 8:
            cout << "Enter pos: ";
            cin >> pos;
            cout << "Enter val: ";
            cin >> val;
            Sl.UpdateList(pos, val);
            break;
        case 9:
            cout << Sl.CountList();
            break;
        case 10:
            Sl.DisplayList();
            break;
        default:
            cout << "\nInvalid Input!!" << endl;
            break;
        }
    } while (ch != 0);

    return 0;

};

File: 01. DataStructures/01. Linked List/3. Circular-Linked-List/circular-linked-list.js
// insert at last 
// reverse list
// print list
// insert at first
// insert at position
//  delete at first
// delete at last
//delete at position

//update list
// count list size


class Node {
   constructor(val, next) {
      this.val = val;
      this.next = next;
   }
}

class circularLinkedList {
    constructor() {
        this.head = null;
        this.size = 0;
    }

    mysize() {
        return this.size;
    }
    insertAtFirstNode(data) {
        this.head = new Node(data);
        this.head.next = this.head;
        this.size++;
    }
    insertAtLast(data) {

        if(!this.head) {
            this.insertAtFirstNode(data);
            return;
        }
        let currunt = this.head;

        while(currunt.next !== this.head) {
            currunt =  currunt.next;
        } 
        currunt.next = new Node(data);
        currunt.next.next = this.head;
        this.size++;
    }

    insertAtFirst(data) {
    //getting the last node 
    let currunt = this.head;
    while(currunt && currunt.next !== this.head) {
        currunt = currunt.next;
    }
    this.head = new Node(data, this.head);
    if(currunt) {
        currunt.next = this.head;
    } else {
        this.head.next = this.head;
    }
    this.size++;
    }

    insertAtPosition1Based(data, pos) {
        if(pos > this.size || pos < 1) return 'enter valid position';
        if(pos == 1) return this.insertAtFirst(data);
        if(pos == this.size + 1) return this.insertAtLast(data);
        let count = 1;
        let currunt = this.head;
        while(currunt.next !== this.head && count < (pos - 1)) {
            currunt = currunt.next;
            count++;
        }
        currunt.next = new Node(data, currunt.next);
        this.size++;
    }
        
    deleteAtFirst() {
       // get the tail node;
       if(this.size == 1) return this.head = null;
       let currunt = this.head;
       while(currunt && currunt.next !== this.head) {
           currunt = currunt.next;
       }
       this.head = this.head.next;
       currunt.next = this.head;       
       this.size--;
    }
    deleteAtLast() {
       if(this.size == 1) return this.head = null;
       let currunt = this.head;
       let count = 1;
       while(count < this.size - 1) {
           currunt = currunt.next;
           count++;
       }
       currunt.next = this.head;
       this.size--;
    }

    deleteAtPosition(pos) {
        if(pos == 1) return this.deleteAtFirst();
        if(pos == this.size) return this.deleteAtLast();
        
        let currunt = this.head;
        let count = 1;
        while(count < pos - 1) {
            currunt = currunt.next;
            count++;
        }
            // 2->3 2->1->3->4 6
        currunt.next = currunt.next.next;
        this.size--;
    }

    printList() {
        if(!this.head) return 'nothing in the list';
        let output = '';
        let currunt = this.head;
        // this is for the first node only.
        output += currunt.val + "->";
        currunt = currunt.next;

        while(currunt !== this.head) {
            output += currunt.val + '->';
            currunt = currunt.next;
        }

        return output;
    }
    // reverse list
    reverseCList() {
        let currunt = this.head;
        let pre = null;
        while(currunt.next !== this.head) {
            const temp = currunt.next;
            currunt.next = pre;
            pre = currunt;
            currunt = temp;
        }
        // main code for circular 
        let temp = currunt.next;
        currunt.next = pre;
        temp.next = currunt;
        // changing the ref of head node
        this.head  = currunt;
    }
}

const cll = new circularLinkedList();

cll.insertAtFirst(6);
cll.insertAtFirst(3);
cll.insertAtFirst(1);
cll.insertAtFirst(29);
cll.insertAtFirst(9);
cll.insertAtFirst(3);
console.log(cll.printList());
// cll.deleteAtPosition(1);
// cll.insertAtLast(2);
// cll.insertAtLast(3);
// cll.insertAtLast(4);
// cll.insertAtLast(5);
// cll.insertAtFirst(6);
// cll.reverseCList();
console.log(cll.printList());
